## CREDIT CARD FRAUD Detection  
  
Importing Libraries

In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

Loadig dataset : 

In [2]:
cc_data = pd.read_csv(r"C:\Users\randa\Purnanand\Projects\Credit Card Fraud\creditcard.csv")

In [3]:
cc_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
cc_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
cc_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

#### We can see that the data is highly unbalanced
0 => Normal Transaction  
1 => Fraudulent Transaction  
#### For analysis we will seperate our data into Legit and Fraud

In [6]:
legit = cc_data[cc_data.Class==0]
fraud = cc_data[cc_data.Class==1]

In [7]:
legit.shape,fraud.shape

((284315, 31), (492, 31))

In [8]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [9]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [10]:
cc_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

### Undersampling
**Build a sample dataset containing similar distribution of Legit transactions and Fraud transactions.**  
  
Number of Fraud transactions = 492  
so, we'll select 492 Legit transactions to make them equal  
As the Data was imbalanced

In [11]:
legit_sample = legit.sample(n=492)

Concatinating the two DataFrames : legit_sample and fraud

In [12]:
new_df = pd.concat([legit_sample, fraud], axis=0)

In [13]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
6689      8309.0 -1.381561 -0.283637  3.697714 -0.303159 -0.521208  0.933298   
114493   73508.0 -0.738367  1.242289  2.134420  0.868145  0.057421 -0.421406   
179631  124168.0  0.444175  0.000043 -0.487265 -0.582484  0.290488  1.426646   
272368  165050.0 -1.152174 -0.915950  0.595726 -2.696335 -0.909675 -1.045305   
6360      7578.0 -0.327810  1.049523  1.809870  1.036585  0.344085  0.373102   

              V7        V8        V9  ...       V21       V22       V23  \
6689   -0.935523  0.448158  2.842411  ... -0.091069  0.686189 -0.442119   
114493  0.733229 -0.084135 -0.907133  ...  0.096063  0.305370 -0.222931   
179631  0.070464  0.138651 -1.721134  ...  0.737737  2.340209 -0.032424   
272368 -0.498195  0.356798 -1.952584  ... -0.006070  0.024704  0.072015   
6360    0.590811 -0.118552  0.704920  ... -0.192627 -0.059591 -0.142747   

             V24       V25       V26       V27       V28  Amount  Class  
6689    0.290457  0.138268  1.122174  0.031293 -0.142384   10.23      0  
114493  0.623270  0.245839 -0.325008 -0.031179  0.086820    4.99      0  
179631 -1.522003 -1.381636  0.413007  0.262067  0.235037   80.75      0  
272368 -0.153984 -0.082207 -0.268215  0.238720  0.049768   91.81      0  
6360   -0.427798 -0.255274 -0.423469  0.213196 -0.030252   36.00      0  

[5 rows x 31 columns]

In [14]:
new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [15]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [16]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      98558.916667  0.136160 -0.039934 -0.035681 -0.082345 -0.108808   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.066256  0.045881  0.040763  0.032302  ... -0.005850  0.012988   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.023212 -0.021402  0.023774  0.019147  0.012741  0.014574 -0.013860   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       94.448272  
1      122.211321  

[2 rows x 30 columns]

#### Splitting data into Features & Targets

In [17]:
X = new_df.drop(columns='Class', axis=1)
Y = new_df['Class']

In [18]:
X.head()

Time        V1        V2        V3        V4        V5        V6  \
6689      8309.0 -1.381561 -0.283637  3.697714 -0.303159 -0.521208  0.933298   
114493   73508.0 -0.738367  1.242289  2.134420  0.868145  0.057421 -0.421406   
179631  124168.0  0.444175  0.000043 -0.487265 -0.582484  0.290488  1.426646   
272368  165050.0 -1.152174 -0.915950  0.595726 -2.696335 -0.909675 -1.045305   
6360      7578.0 -0.327810  1.049523  1.809870  1.036585  0.344085  0.373102   

              V7        V8        V9  ...       V20       V21       V22  \
6689   -0.935523  0.448158  2.842411  ... -0.049470 -0.091069  0.686189   
114493  0.733229 -0.084135 -0.907133  ...  0.048730  0.096063  0.305370   
179631  0.070464  0.138651 -1.721134  ...  0.293870  0.737737  2.340209   
272368 -0.498195  0.356798 -1.952584  ... -0.062624 -0.006070  0.024704   
6360    0.590811 -0.118552  0.704920  ...  0.332835 -0.192627 -0.059591   

             V23       V24       V25       V26       V27       V28  Amount  
6689   -0.442119  0.290457  0.138268  1.122174  0.031293 -0.142384   10.23  
114493 -0.222931  0.623270  0.245839 -0.325008 -0.031179  0.086820    4.99  
179631 -0.032424 -1.522003 -1.381636  0.413007  0.262067  0.235037   80.75  
272368  0.072015 -0.153984 -0.082207 -0.268215  0.238720  0.049768   91.81  
6360   -0.142747 -0.427798 -0.255274 -0.423469  0.213196 -0.030252   36.00  

[5 rows x 30 columns]

In [19]:
Y

6689      0
114493    0
179631    0
272368    0
6360      0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

#### Splitting the data into Train and Test

In [20]:
from sklearn.model_selection import train_test_split

In [21]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2, stratify=Y, random_state=2)

In [22]:
X.shape, X_train.shape, X_test.shape

((984, 30), (787, 30), (197, 30))

In [23]:
Y.shape, Y_train.shape, Y_test.shape

((984,), (787,), (197,))

## MODEL TRAINING 
We will we applying the following Algoritms :  
* 1. Logistic Regression
* 2. SVM
* 3. Random Forest
* 4. Bagging
* 5. XGBoost

In [24]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

from sklearn.metrics import accuracy_score

### 1. LOGISTIC REGRESSION

In [25]:
model = LogisticRegression()

TRAINING OUR MODEL : 

In [26]:
# training the Logistic Regression Model with Training Data

model.fit(X_train, Y_train)

LogisticRegression()

#### Model Evaluation

PREDICTION : 

In [27]:
X_train_prediction = model.predict(X_train)

In [28]:
X_test_prediction = model.predict(X_test)

**Accuracy Score :**

TRAINING DATA  : 

In [29]:
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [30]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9529860228716646


TEST DATA : 

In [31]:
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [32]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.9441624365482234


In [33]:
all_accuracy = {}

In [34]:
all_accuracy['Logistic']=test_data_accuracy

### 2. SUPPORT VECTOR MACHINE

In [35]:
model = SVC(kernel="linear")

TRAINING OUR MODEL : 

In [36]:
# training the Logistic Regression Model with Training Data

model.fit(X_train, Y_train)

SVC(kernel='linear')

#### Model Evaluation

PREDICTION : 

In [37]:
X_train_prediction = model.predict(X_train)

In [38]:
X_test_prediction = model.predict(X_test)

**Accuracy Score :**

TRAINING DATA  : 

In [39]:
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [40]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9085133418043202


TEST DATA : 

In [41]:
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [42]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.9086294416243654


In [43]:
all_accuracy['SVM']=test_data_accuracy

### 3. RANDOM FOREST CLASSIFIER

In [61]:
model = RandomForestClassifier()

TRAINING OUR MODEL : 

In [62]:
# training the Logistic Regression Model with Training Data

model.fit(X_train, Y_train)

RandomForestClassifier()

#### Model Evaluation

PREDICTION : 

In [63]:
X_train_prediction = model.predict(X_train)

In [64]:
X_test_prediction = model.predict(X_test)

**Accuracy Score :**

TRAINING DATA  : 

In [65]:
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [66]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  1.0


TEST DATA : 

In [67]:
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [68]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.9187817258883249


In [69]:
all_accuracy['RandomForest']=test_data_accuracy

### 4. Bagging

In [70]:
    from sklearn.ensemble import BaggingClassifier
    model = BaggingClassifier(random_state = 100)
    model.fit(X_train, Y_train)
    X_train_prediction = model.predict(X_train)
    X_test_prediction = model.predict(X_test)
    test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
    print(f'Accuracy score on Test Data  : ', test_data_accuracy)

Accuracy score on Test Data  :  0.9035532994923858


In [71]:
all_accuracy['Bagging']=test_data_accuracy

### 5. XGBoost

In [72]:
Y_train = Y_train.astype('category').cat.codes
Y_train.value_counts()
Y_train.astype('category')
Y_test = Y_test.astype('category').cat.codes
Y_test.astype('category')

247995    1
147541    0
54095     0
64604     0
15451     1
         ..
108258    1
12108     1
99506     1
74496     1
20049     0
Length: 197, dtype: category
Categories (2, int64): [0, 1]

In [73]:
import xgboost
from xgboost import XGBClassifier
model_Xgboost = XGBClassifier()
model_Xgboost.fit(X_train, Y_train)
y_pred_XGBoost = model_Xgboost.predict(X_test)
 
print("Accuracy is:",accuracy_score(Y_test,y_pred_XGBoost))


Accuracy is: 0.9137055837563451


In [74]:
all_accuracy['XGBoost']=test_data_accuracy

In [75]:
Accuracy = pd.DataFrame([all_accuracy]).T
Accuracy = Accuracy.rename(columns={0:'Accuracy'})

In [76]:
Accuracy

Accuracy
Logistic      0.944162
SVM           0.908629
RandomForest  0.918782
Bagging       0.903553
XGBoost       0.903553

We can see that we that we get the most accuracy score for **Logistic Regression Algorithm** i.e. 94.416%